forte/predictor.py

```diff
@@ -0,0 +1,185 @@
+# Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import Dict, List, Callable
+import itertools
+from torch.nn import Module
+from forte.data.types import DATA_INPUT
+from forte.data.data_pack import DataPack
+from forte.data.ontology import Annotation
+from forte.processors.base.base_processor import BaseProcessor
+from forte.process_manager import ProcessJobStatus
+
+
+class Batcher:
+    '''This class will create a pool of pack, instance and feature. When
+    the pool is filled with a batch size of elements. It will generate them
+    in a batch. The batched data will be pass to the prediction function in the
+    Predictor. Note that the extract, convert process is done during
+    this batching process.
+    '''
+    def __init__(self, batch_size: int, feature_resource: Dict,
+                cross_pack=True):
+        # pylint: disable=line-too-long
+        '''Feature_resource is prodcued from train pipeline.
+        An example looks like
+        {
+        'scope': ft.onto.base_ontology.Sentence,
+        'schemes': {
+            'text_tag': {
+                'extractor': <forte.data.extractor.extractor.TextExtractor>,
+                'converter': <forte.data.extractor.converter.Converter>,
+                'type': DATA_INPUT
+            },
+
+            'char_tag': {
+                'extractor': <forte.data.extractor.extractor.CharExtractor>,
+                'converter': <forte.data.extractor.converter.Converter>,
+                'type': DATA_INPUT
+            },
+
+            'ner_tag': {
+                'extractor': <forte.data.extractor.extractor.AnnotationSeqExtractor>,
+                'converter': <forte.data.extractor.converter.Converter>,
+                'type': DATA_OUTPUT,
+                'unpadder': <forte.data.extractor.converter.SameLengthUnpadder>}
+            }
+        }
+        '''
+        self.batch_size = batch_size
+        self.feature_resource = feature_resource
+        self.cross_pack = cross_pack
+        self.pack_pools = []
+        self.instance_pools = []
+        self.features_pools = []
+
+    def current_pack_len(self) -> int:
+        return len(set(self.pack_pools))
```

Comment:
Why do we need to do the`set` step here? What if the user really want to pass in a pack twice?